src/System.Private.CoreLib/src/System/Threading/ClrThreadPool.WaitThread.cs

```diff
@@ -0,0 +1,281 @@
+                        if (_registeredWaitHandles[i] == null)
+                        {
+                            _numUserWaits = i - 1;
+                            break;
+                        }
+                    }
+                }
+
+                // Recalculate Timeout
+                int timeout = Timeout.Infinite;
+                for (int i = 0; i < _numUserWaits; i++)
+                {
+                    if (timeout == Timeout.Infinite)
+                    {
+                        timeout = _registeredWaitHandles[i].Timeout;
+                    }
+                    else
+                    {
+                        timeout = Math.Min(_registeredWaitHandles[i].Timeout, timeout);
+                    }
+                }
+                _currentTimeout = timeout;
+                _registeredHandlesLock.Release();
+                _removesLock.Release();
+            }
+
+            private void QueueWaitCompletion(RegisteredWaitHandle registeredHandle, bool timedOut)
+            {
+                ThreadPool.QueueUserWorkItem(CompleteWait, new CompletedWaitHandle(registeredHandle, timedOut));
+            }
+
+            private void CompleteWait(object state)
+            {
+                CompletedWaitHandle handle = (CompletedWaitHandle)state;
+                handle.CompletedHandle.CanUnregister.Reset();
+                handle.CompletedHandle.PerformCallback(handle.TimedOut);
+                handle.CompletedHandle.CanUnregister.Set();
+                if (!handle.CompletedHandle.Repeating)
+                {
+                    QueueOrExecuteUnregisterWait(handle.CompletedHandle);
+                }
+            }
+
+            public bool RegisterWaitHandle(RegisteredWaitHandle handle)
+            {
+                StartWaitThreadIfNotStarted();
+
+                if (_numUserWaits == WaitHandle.MaxWaitHandles - 1)
+                {
+                    return false;
+                }
+
+                AddWaitHandleForNextWait(handle);
+                handle.WaitThread = this;
+                _changeHandlesEvent.Set();
+                return true;
+            }
+
+            private void AddWaitHandleForNextWait(RegisteredWaitHandle handle)
+            {
+                _registeredHandlesLock.Acquire();
+
+                _registeredWaitHandles[_numUserWaits] = handle;
+                _waitHandles[_numUserWaits + 1] = handle.Handle;
+                if (_currentTimeout == Timeout.Infinite)
+                {
+                    _currentTimeout = handle.Timeout;
+                }
+                else
+                {
+                    _currentTimeout = Math.Min(_currentTimeout, handle.Timeout);
+                }
+                _numUserWaits++;
+                _registeredHandlesLock.Release();
+            }
+
+            private void StartWaitThreadIfNotStarted()
+            {
+                _waitThreadStartedLock.Acquire();
+                if (!_waitThreadStarted)
```

Comment:
For better performance, this condition should be checked before acquiring the lock and then re-checked again with lock acquired.
In addition, I do not see where _waitThreadStarted is set to true.